docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,679 @@
+The memory manager must be initialized in first ECALL (ECMD_INIT_ENCLAVE in
+Intel SGX SDK) before any other clients can use it. Therefore, code and data
+of the memory manager will be part of initial enclave image that are loaded
+with EADD before EINIT, and as a part of the trusted runtime.
+
+The trusted runtime should enumerate all initial committed regions (code,
+data, heap, stack, TCS, SSA), and call the EMM internal APIs to setup
+initial entries in the EMA list to track existing regions and mark some
+of them not modifiable by EMM public APIs. Runtime also ensures there is
+enough reserved space on heap for EMM to create the initial EMA list and
+the entries. Once initialized, the memory manager can reserve its own
+space for future expansion of the EMA list, and special EMAs to hold
+EMA objects. To keep it simple, the expansion can be done eagerly: commit
+more pages for EMA list once unused committed space in the EMA List Region
+below certain threshold.
+
+Alternative option: At build time, the enclave signing tool can precalculate
+and fill in EMA entries that holds info of initial regions to be committed by
+EADD during enclave load. The calculated start addresses in these EMAs can be
+relative to enclave secs->base. Runtime can patch those entries at
+initialization by adding secs->base. The EMM can directly use those EMAs as
+the initial entries of the EMA list. It only needs to reserve and commit
+a number of additional pages for future EMA list expansion.
+
+
+### EMM Private APIs for Trusted Runtimes
+These private APIs can be used by the trusted runtime to reserve and allocate
+regions not accessible from public APIs. They have the identical signature
+as the public API counterparts and replace "sgx_mm_" prefix with "ema_" prefix.
+The main difference is that the private ema_alloc allows an extra flag
+EMA_SYSTEM passed in.
+
+```
+#define EMA_SYSTEM EMA_ALLOC_FLAGS(0x40) /* EMA reserved by system */
+
+/**
+ * Same as sgx_mm_alloc, EMA_SYSTEM can be OR'ed with flags to indicate
+ * that the EMA can not be modified thru public APIs
+ */
+int ema_alloc(void *addr, size_t length, int flags,
+              enclave_fault_handler_t handler,
+              void *handler_private,
+              void **out_addr);
+
+int ema_dealloc(void *addr, size_t length);
+
+int ema_uncommit(void *addr, size_t length);
+
+int ema_commit(void *addr, size_t length);
+
+int ema_commit_data(void *addr, size_t length, uint8_t *data, int prot);
+
+int ema_modify_ex(void *addr, size_t length, int prot, int type);
+
+```
+
+Internal APIs and Structures
+-------------------------------------
+
+Following are internal functions and structures to be used by EMM implementation.
+They can evolve over time, shown here for reference only.
+
+### Enclave Memory Area (EMA) struct
+
+Each enclave has a global doubly linked EMA list to keep track of all dynamically
+allocated regions in enclave address space (ELRANGE).
+
+```
+typedef BitArray;
```

Comment:
what is BitArray?  should it be `sgx_bit_array_t`?

Reply:
will remove this definition. It is implementation detail and  the comments for eaccept_map should be enough for design doc.